scrapy_google_sheets_exporter/gsheets_exporter.py

```diff
@@ -0,0 +1,79 @@
+import csv
+import logging
+from io import StringIO
+
+import gspread
+from scrapy.exceptions import NotConfigured
+from scrapy.extensions.feedexport import BlockingFeedStorage, build_storage
+
+logger = logging.getLogger(__name__)
+
+
+class GoogleSheetsFeedStorage(BlockingFeedStorage):
+    def __init__(self, uri, credentials, *, feed_options=None):
+        if credentials:
+            self.gc = gspread.service_account_from_dict(credentials)
+        else:
+            raise NotConfigured(
+                "Must specify GOOGLE_CREDENTIALS (dict) in the spider settings."
+            )
+
+        self.spreadsheet_key, self.sheet_name = self.parse_gsheets_uri(uri)
+        self.spreadsheet = self.gc.open_by_key(self.spreadsheet_key)
+        self.sheet = self.spreadsheet.worksheet(self.sheet_name)
+        self.feed_options = feed_options or {}
+        self.append_mode = self.feed_options.get("append_mode", False)
+        self.fields = self.feed_options.get("fields", [])
+        self.format = self.feed_options.get("format", "csv")
+
+        if self.format != "csv":
+            raise NotConfigured(
+                "This feed exporter only supports csv format. "
+                f"Please update the FEEDS settings by replacing {self.format} format."
+            )
+
+    @classmethod
+    def from_crawler(cls, crawler, uri, *, feed_options=None):
+        return build_storage(
+            cls,
+            uri,
+            credentials=crawler.settings["GOOGLE_CREDENTIALS"],
+            feed_options=feed_options,
+        )
+
+    def _store_in_thread(self, file):
+        file.seek(0)
+        csv_text = file.read().decode("utf-8")
+        csv_reader = csv.reader(StringIO(csv_text))
```

Comment:
Can’t you pass `file` to `csv.reader` directly somehow, and also iterate CSV rows and add them to the sheet during iteration instead of reading all records first? (to avoid reading the whole CSV file into memory)

Reply:
I agree with @Gallaecio . Loading the entire file in memory may be problematic for larger files.

Reply:
Yes, I can pass the file to the csv reader directly if I use TextIOWrapper, I've made the change now, please take a look, thanks!